feat: make 'preinsert' work with 'autocomplete'

girishji opened the pull request:
This change extends Insert mode autocompletion so that "preinsert" behavior of 'completeopt' also
works when 'autocomplete' is enabled. The behavior is slightly different.

Try: `:set autocomplete completeopt=preinsert`. It's quite useful.

See `:help 'cot'` for more details.

Also, added example setup for `autocomplete`. See `:h ins-autocompletion-example`.

girishji committed:
feat: make 'preinsert' work with 'autocomplete'

This change extends Insert mode autocompletion so that 'preinsert' also
works when 'autocomplete' is enabled.

Try: `:set ac cot=preinsert`

See `:help 'cot'` for more details.

habamax commented:
Looks like `hl-ComplMatchIns` is not defined in vim/colorschemes. Is it a new one or I completely overlooked its existence?

<img width="375" height="208" alt="image" src="https://github.com/user-attachments/assets/ac94bc73-4d67-4c74-a937-16592e53cbed" />

girishji commented:
It was already there. I just noticed it was undefined, so I linked it to the `Added` highlight group.

chrisbra commented:
Thanks, but this seems to break a few screen-dump tests

girishji commented:
> Thanks, but this seems to break a few screen-dump tests


Thanks, just noticed this. The failures are caused by the highlight group `ComplMatchIns`, which I had linked incorrectly to another group. That was a mistake. `ComplMatchIns` isn’t limited to *preinsert* -- it also highlights the postfix part of the text when selecting from the popup menu.

I am not sure if it would make sense to have a separate highlight group from *preinsert*. For now, I reverted the changes so as not to hold up this PR. 

@habamax I think `hl-ComplMatchIns` can’t really be used in any colorscheme, because it highlights every match you pick from the menu. That feels quite unusual.

girishji committed:
Add hl-PreInsert

Added a new highlight group since preinsert is a different function than what
hl-ComplMatchIns serves.

girishji commented:
Added a new highlight group `hl-PreInsert`.  Using `hl-ComplMatchIns` for both purposes prevented setting distinct default colors for pre-inserted text.  Currently applies only to 'autocomplete'.

girishji committed:
Improve checks

M  src/edit.c
M  src/insexpand.c

girishji committed:
Improve doc

M  runtime/doc/insert.txt

chrisbra commented:
It should be `ComplMatchPreIns` no? We can rename it

girishji commented:
I wasn’t aware there was a convention. I chose what felt the most intuitive and readable. When I first saw `hl-ComplMatchIns`, it looked like a jumble of letters I couldn’t immediately connect to anything. If a convention is already established, I don’t mind following it, but it feels odd to replace a meaningful tag with something that looks cryptic.

chrisbra commented:
> The message could have been a tiny bit more polite.

agree.

> if a convention is already established, I don’t mind following it, but it feels odd to replace a meaningful tag with something that looks cryptic.

Well, we only have a single highlight group for completion: `ComplMatchIns` so it seems we have a precedent for other related highlighting groups. So if there is a chance we will have more additional related highlighting groups related to completion feature, I think this would make sense to have a consistent naming.

girishji commented:
The doc is incorrect:

```
ComplMatchIns   Matched text of the currently inserted completion.
```

This highlight group applies to the _unmatched_ portion of the completion (the suffix). Consider renaming it to something like `InsertedSuffix`, `CompletionSuffix`, `PostfixText`, etc. Current name conveys the opposite meaning of what it is doing.

habamax commented:
According to the document you have recently authored, it is not too late :)

FWIW, it is not in any of built-in colorschemes.

habamax commented:
> in that case what would be a good naming convention? Something that would also apply for PreInsert?

Should we have a prefix for the group of 2 (as of now) highlight groups?

- CompletePreInsert, CoPreInsert, CtePreInsert, ...
- CompleteMatchInsert, CoMatchInsert, CteMatchInsert, ... or CompleteSuffix, CompleteMatchSuffix, CoMatchSuffix, CteMatchSuffix

Shane-XB-Qian commented:
> According to the document you have recently authored, it is not too late :)
> 
> FWIW, it is not in any of built-in colorschemes.

but it maybe in some user's vimrc already, (supposed you meant the 'policy' new added into develop.txt),
as usually in the past it was allowed maximal 1 week, minimal 1 day, i may think the 'policy' should keep that way.

now this 'CompMatchIns' had been over many weeks, i think better be keep it is now;
as for 'preinsert', how about 'ComplPreIns' or 'ComplMatchPreIns'?

-- 
shane.xb.qian
